File: Python/Trabalho_POO_parte_2/Motocicleta.py
from Veiculo_Motorizado import *
from Veiculo import *
from Interface_IPVA import *

class Motocicleta(Veiculo_Motorizado, Interface_IPVA):
    # metodo final(isso é feito implicitaente, linguagem ja interpleta que é uma variavel que não pode ser alterada)
    __gasto_motocicleta: float = 0.25
    #contrut
    def __init__(self, id: int, rodas: int):
        super().__init__(id, rodas)
        super().set_valor_ipva(self.calcular_ipva())
        self.__tipo = "M"
        print(f"Motocicleta com ID {self.get_id()} foi construida")

    # metodo get
    def get_tipo(self) -> int:
        return self.__tipo

    # funções abstrata herdada pelo pai veiculo 
    def mover(self):
        if self.get_ipva():
            if self.get_combustivel() >= self.__gasto_motocicleta:
                for roda in self.get_rodas():
                    if roda.verificar_pneu():
                        self.consumir(self.__gasto_motocicleta)
                        new_distancia = self.get_distancia_percorrida() + 2
                        self.set_distancia_percorrida(new_distancia)
                        print(
                            f"Motocicleta - ID:{self.get_id()} andou  bloco +3 e está com {self.get_combustivel():.2f}litros de gasolina")
                        return
                print(
                    f"Motocicleta com ID {self.get_id()} não possui todos os pneus calibrados")
            else:
                print(
                    f"Motocicleta com ID {self.get_id()} não possui combustivel suficiente")
        else:
            print(f"Motocicleta com ID {self.get_id()} não está com ipva em dia")

    def desenhar(self):
        print(" "*self.get_distancia_percorrida() + "  , _oo")
        print(" "*self.get_distancia_percorrida() + ".-/c-//: :")
        print(" "*self.get_distancia_percorrida() + "(_)'=='(_)")
    
    def calcular_ipva(self):
        self.valor =  self.VALOR_BASE * self.MOTOCICLETA

File: C++/Trabalho_POO_parte_2/Motocicleta.cpp
#include "Motocicleta.h"


Motocicleta::Motocicleta(int id, int rodas) : Veiculo_Motorizado(id, rodas), tipo("M")
{
    cout << "Motocicleta com ID " << get_id() << " foi construída" << endl;
}

 string Motocicleta::get_tipo() const
{
    return tipo;
}

 void mover() override
{
    if (get_ipva())
    {
        if (get_combustivel() >= gasto_motocicleta)
        {
            if (verificar_pneu())
            {
                consumir(gasto_motocicleta);
                int new_distancia = get_distancia_percorrida() + 2;
                set_distancia_percorrida(new_distancia);
                cout << "Motocicleta - ID:" << get_id() << " andou +2 blocos e está com " << get_combustivel() << " litros de gasolina" << endl;
            }
            else
                cout << "Motocicleta com ID " << get_id() << " não possui todos os pneus calibrados" << endl;
        }
        else
            cout << "Motocicleta com ID " << get_id() << " não possui combustível suficiente" << endl;
    }
    else
        cout << "Motocicleta com ID " << get_id() << " não está com o IPVA em dia" << endl;
}

 void desenhar() override
{
    string espacos(get_distancia_percorrida(), ' ');
    cout << espacos << "  , _oo" << endl;
    cout << espacos << ".-/c-//: :" << endl;
    cout << espacos << "(_)'=='(_)" << endl;
}